src/storage/mvcc/txn.rs

```diff
@@ -203,9 +219,92 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(())
     }
 
+    pub fn pessimistic_lock(
+        &mut self,
+        key: Key,
+        primary: &[u8],
+        for_update_ts: u64,
+        options: &Options,
+    ) -> Result<()> {
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.to_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            MVCC_DUPLICATE_CMD_COUNTER_VEC.pessimistic_lock.inc();
+            return Ok(());
+        }
+
+        if let Some((commit, write)) = self.reader.seek_write(&key, u64::max_value())? {
+            // The isolation level of pessimistic transactions is RC. `for_update_ts` is
+            // the commit_ts of the data this transaction read. If exists a commit version
+            // whose commit timestamp is larger than current `for_update_ts`, the
+            // transaction should retry to get the latest data.
+            //
+            // If the commit timestamp is equal to transaction's start timestamp, the
+            // transaction is already rollbacked.
+            if commit > for_update_ts || commit == self.start_ts {
+                MVCC_CONFLICT_COUNTER.pessimistic_lock_conflict.inc();
+                return Err(Error::WriteConflict {
+                    start_ts: for_update_ts,
+                    conflict_start_ts: write.start_ts,
+                    conflict_commit_ts: commit,
+                    key: key.to_raw()?,
+                    primary: primary.to_vec(),
+                });
+            }
+
+            // Handle rollback.
+            // If `commit` we seek is already before `start_ts`, the rollback must not exist.
+            if commit > self.start_ts {
```

Comment:
When there is a commit_ts > current start_ts, also can aquire pessimistic lock?

Reply:
pessimistic lock checks previous written transactions by `for_update_ts` rather than `start_ts`.